experiment/simulation/js/flipflop.js

```diff
@@ -0,0 +1,780 @@
+            }
+            else if (s && !r) {
+                this.q = true;
+                this.qbar = false;
+            }
+            else if (s && r) {
+                this.q = true;
+                this.qbar = true;
+            }
+        }
+    }
+}
+
+
+export function addRSFlipFlop(x, y) {
+    const ff = new RSFlipFlop();
+    ff.registerComponent("working-area", x, y);
+}
+
+window.addRSFlipFlop = addRSFlipFlop;
+
+export function getOutputRS(gate, pos) {
+    if (pos === "Q") {
+        return gate.q;
+    }
+    else if (pos === "Q'") {
+        return gate.qbar;
+    }
+    // But if the gate is not an FA, but an input bit, then return the value of the input
+    else {
+        return gate.output
+    }
+}
+
+// done checking
+export function getResultRS(ff) {
+    // check if flipflop type is Gate object
+    if (ff.constructor.name === "Gate") {
+        return;
+    }
+
+    if (getOutputRS(ff.r[0], ff.r[1]) != null && getOutputRS(ff.s[0], ff.s[1]) != null && getOutputRS(ff.clk[0], ff.clk[1]) != null) {
+        ff.generateOutput();
+    }
+    return;
+}
+
+
+// done checking
+export function checkConnectionsRS() {
+    let correctConnection = true;
+    for (let ffID in flipFlops) {
+        const gate = flipFlops[ffID];
+        // For Full Adder objects
+        // Check if all the outputs are connected
+        if (!gate.qIsConnected) {
+            correctConnection = false;
+            break;
+        }
+        if (!gate.qbarIsConnected) {
+            correctConnection = false;
+            break;
+        }
+        // Check if all the inputs are connected
+        if (gate.r == null || gate.r.length === 0) {
+            correctConnection = false;
+            break;
+        }
+        if (gate.s == null || gate.s.length === 0) {
+            correctConnection = false;
+            break;
+        }
+        if (gate.clk == null || gate.clk.length === 0) {
+            correctConnection = false;
+            break;
+        }
+    }
+    for (let gateId in gates) {
+        const gate = gates[gateId];
+        if (gate.isInput) {
+            if (!gate.isConnected) {
+                correctConnection = false;
+                break;
+            }
+        }
+        else if (gate.isOutput) {
+            if (gate.inputs.length === 0) {
+                correctConnection = false;
+                break;
+            }
+        }
+        else {
+            if (gate.inputPoints.length != gate.inputs.length) {
+                correctConnection = false;
+            }
+            else if (!gate.isConnected && !gate.isOutput) {
+                correctConnection = false;
+            }
+        }
+    }
+
+    if (correctConnection) {
+        return true;
+    }
+    else {
+        // printErrors("Connections are not correct\n");
+        // alert("Connections are not correct");
+        return false;
+    }
+}
+
+export function simulateFFRS() {
+    for (let ffID in flipFlops) {

```

Comment:
Please ensure that you need to use the for..in loop and not the for..of loop

Reply:
using for..of in those places instead of for..in throws error saying that the list is not iterable, so in such places we have used for..in only